Rework output function to interpolate between min|max limits.

FreeBear-nc opened the pull request:
Never been happy with the way output was calculated. arthurrump's routine would overshoot max_value.
My original revision clamped the output to min|max values whilst retaining the (faulty ?) interpolation.
This commit correctly (in my opinion) interpolates output between min and max values using library functions.

Downside is anyone switching to this code will have to retune their PID parameters. But it is a price I'm willing to pay.

FreeBear-nc committed:
The original output function used "min_value + state * (max_value - min_value)"
 to normalise the value - Basically, a lerp() function, but state needs to be
 in the 0-1 range to interpolate properly.

Use remap() to limit the range to 0-1 - If state < min_value, the result will
be negative, and if state > max_value, it still needs clamping to avoid overshoot.

FreeBear-nc committed:
Merge branch 'olegtarasov:main' into main

olegtarasov commented:
To be honest, I don't think this is the right approach. Your code assumes that PID output values are in range (`min_value`; `max_value`). Obviously, PID can output values in any range, which is controlled by its parameters. But a quick look that I took at PID Autotuner that is built into ESPHome, suggests that it tunes PID parameters so that its output is in range (-1; 1). If we apply your change, people would get wrong parameters from the built-in autotuner.
Also, we don't have any mechanism to notify people who already use the component about this change. And since most people just follow the main branch, they will suddenly get broken PID configuration, which will be very hard to debug.
Bearing in mind these two issues, I honestly think we should not introduce this change. It doesn't change anything conceptually (we still need to clamp, we just change our assumption about PID output range), but introduces breaking changes to setups that already work.

FreeBear-nc commented:
>  But a quick look that I took at PID Autotuner that is built into ESPHome, suggests that it tunes PID parameters so that its output is in range (-1; 1). 

Unless I'm mistaken, the -1:1 range is just for the integral term (Ki). That doesn't prevent integral wind-up, nor does it clamp the output.
If we agree that state (output) should be clamped to min|max values, using remap() and lerp() just smooths the curve  and avoids any big jumps in output. Using PID Autotune to get a new set of Kp, Ki, and Kp parameters is no big big deal in my mind, so doesn't really introduce a major breakage. Changes to sensor names plus a few missing ones has meant that migrating my viessmann.yaml needed quite a few changes. Breakages were expected, and any new ones could (should ?) be documented in the README.

FreeBear-nc committed:
Remove the remap() call - We can debate it another day.
In the meantime, lerp() does exactly the same as the code it replaces so may
 as well use the helpers provided.

FreeBear-nc commented:
Removed the remap() call, but retained use of the lerp() function. We can debate the use of remap() another day.
